prov/util/src/util_buf.c

```diff
@@ -45,6 +45,44 @@ enum {
 };
 
 
+static int ofi_bufpool_alloc(struct ofi_bufpool_region *buf_region)
+{
+	int ret;
+	struct ofi_bufpool *pool = buf_region->pool;
+
+	if (pool->attr.flags & OFI_BUFPOOL_HUGEPAGES) {
+		ret = ofi_alloc_hugepage_buf((void **) &buf_region->alloc_region,
+					     pool->alloc_size);
+		if (!ret) {
+			buf_region->flags = OFI_BUFPOOL_HUGEPAGES;
+			return 0;
+		}
+		/* If we can't allocate huge pages, fall back to normal
+		 * allocations for all future attempts.
+		 */
+		pool->attr.flags &= ~OFI_BUFPOOL_HUGEPAGES;
+	}
+
+	if (pool->attr.flags & OFI_BUFPOOL_MMAP) {
```

Comment:
Shouldn't this be an else if? I don't see a case where we will try mmap after trying allocating hugepages...

Reply:
This is on purpose.  I originally wrote this to fall through to mmap if huge page allocation failed, but decided to leave this to follow-up discussion.

The only time we'll ask for huge pages is when fork support is not needed, so in theory it would be safe to just use small pages as before.  However, it may be nice to move the buffers involved with `ibv_reg_mr` into a separate mmap anyway.  This could be accomplished by keeping the fall-through here and specifying `OFI_BUFPOOL_HUGEPAGES | OFI_BUFPOOL_MMAP` in efa (see https://github.com/nzmsv/libfabric/pull/1)

Reply:
Since, its a core change (not just specific to EFA provider), there may be cases where providers would allocate from hugepages or fallback to malloc. I think if provider would like an option to try mmap after failed hugepage allocation then they can use enable both the flags as you suggested.

Reply:
@shefty also agreed (and provided the same feedback) so I am keeping this